addons/crm/models/crm_lead.py

```diff
@@ -595,8 +595,15 @@ def write(self, vals):
         if any(field in ['active', 'stage_id'] for field in vals):
             self._handle_won_lost(vals)
 
+        from_lead_values = dict([lead.id, [lead.stage_id.id, lead.active]] for lead in self)
         write_result = super(Lead, self).write(vals)
 
+        # Recompute the probabilities for a lead if its stage or activation has changed
+        # This allows to take into account the table update in _handle_won_lost
+        for lead in self:
+            if from_lead_values[lead.id][0] != lead.stage_id.id or from_lead_values[lead.id][1] != lead.active:
+                self._compute_probabilities()
```

Comment:
you should make a recordset that contains all the leads that have changed their 'stage' or 'active' value and run compute_probabilities on them only. (as it's not necessary to recompute the rest a priori)

to_recompute = self.env["crm.lead"]
for lead in self:
   if ...:
       to_recompute |= lead
to_recompute.compute_prob()